ERCS/erc-7677.md

```diff
@@ -0,0 +1,296 @@
+
+```json
+{
+  "paymasterAndData": "0x..."
+}
+```
+
+If using entrypoint v0.7:
+
+```json
+{
+  "paymaster": "0x...",
+  "paymasterData": "0x..."
+}
+```
+
+#### `pm_getPaymasterData`
+
+Returns values to be used in paymaster-related fields of a signed user operation. These are not stub values and will be used during user operation submission to a bundler. Similar to `pm_getPaymasterStubData`, accepts an unsigned user operation, entrypoint address, chain id, and a context object.
+
+This method also returns a `sponsor` object with `name` and `icon` fields. The `name` field is the name of the party sponsoring the transaction, and the `icon` field is a URI pointing to an image. Wallet developers MAY choose to display sponsor information to users. The `icon` string MUST be a data URI as defined in [RFC-2397]. The image SHOULD be a square with 96x96px minimum resolution. The image format is RECOMMENDED to be either lossless or vector based such as PNG, WebP or SVG to make the image easy to render on the wallet. Since SVG images can execute Javascript, wallets MUST render SVG images using the `<img>` tag to ensure no untrusted Javascript execution can occur.
+
+##### `pm_getPaymasterData` RPC Specification
+
+```typescript
+// [userOp, entryPoint, chainId, context]
+type GetPaymasterDataParams = [
+  // Below is specific to Entrypoint v0.6 but this API can be used with other entrypoint versions too
+  {
+    sender: `0x${string}`;
+    nonce: `0x${string}`;
+    initCode: `0x${string}`;
+    callData: `0x${string}`;
+    callGasLimit: `0x${string}`;
+    verificationGasLimit: `0x${string}`;
+    preVerificationGas: `0x${string}`;
+    maxFeePerGas: `0x${string}`;
+    maxPriorityFeePerGas: `0x${string}`;
+  }, // userOp
+  `0x${string}`, // Entrypoint
+  `0x${string}`, // Chain ID
+  Record<string, any> // Context
+];
+
+type GetPaymasterDataResult = {
+  sponsor?: { name: string; icon: string } // Sponsor info
+} & Record<string, `0x${string}`>; // Paymaster values
+```
+
+###### `pm_getPaymasterData` Example Parameters
+
+```json
+[
+  {
+    "sender": "0x...",
+    "nonce": "0x...",
+    "initCode": "0x",
+    "callData": "0x...",
+    "callGasLimit": "0x...",
+    "verificationGasLimit": "0x...",
+    "preVerificationGas": "0x...",
+    "maxFeePerGas": "0x...",
+    "maxPriorityFeePerGas": "0x...",
+    "paymasterAndData": "0x...",
+    "signature": "0x..."
+  },
+  "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789",
+  "0x2105",
+  {
+    // Illustrative context field. These should be defined by service providers.
+    "policyId": "962b252c-a726-4a37-8d86-333ce0a07299"
+  }
+]
+```
+
+###### `pm_getPaymasterData` Example Return Value
+
+Paymaster services MUST detect which entrypoint version the account is using and return the correct fields.
+
+For example, if using entrypoint v0.6:
+
+```json
+{
+  "sponsor": {
+    "name": "My App",
+    "icon": "https://..."
+  },
+  "paymasterAndData": "0x..."
+}
+```
+
+If using entrypoint v0.7:
```

Comment:
I think I disagree with having different specs for v0.7 vs v0.6. In both cases paymasterAndData is a concatenation of different values. It easier to just supply the paymasterAndData field so wallet doesn't have to handle.

Reply:
discussed with Wilson offline, these fields should be separate as currently written in the draft. ie paymasterAndData for v0.6 and paymaster / paymasterData for v0.7.

the reasoning is that while these fields are packed onchain (i.e. once user op reaches the entrypoint, it will still be `paymasterAndData`), wallets submit ops to bundlers unpacked, so keeping them separate in the paymaster response would be easiest for wallets to handle.